wgpu-types/src/lib.rs

```diff
@@ -2458,3 +2462,71 @@ pub enum SamplerBorderColor {
     OpaqueBlack,
     OpaqueWhite,
 }
+
+/// Describes how to create a QuerySet.
+#[derive(Clone, Debug)]
+pub struct QuerySetDescriptor {
+    /// Kind of query that this query set should contain.
+    pub ty: QueryType,
+    /// Total count of queries the set contains. Must not be zero.
+    /// Must not be greater than [`QUERY_SET_MAX_QUERIES`].
+    pub count: u32,
+}
+
+/// Type of query contained in a QuerySet.
+#[derive(Copy, Clone, Debug)]
+pub enum QueryType {
+    /// Query returns up to 5 64-bit numbers based on the given flags.
+    ///
+    /// See [`PipelineStatisticsTypes`]'s documentation for more information
+    /// on how they get resolved.
+    ///
+    /// [`Features::PIPELINE_STATISTICS_QUERY`] must be enabled to use this query type.
+    PipelineStatistics(PipelineStatisticsTypes),
```

Comment:
We must be careful here. The upstream API works with a list of types, and it makes it so the implementation has to write down the results according to the list.
In this PR it's a set, however, so what order do the values come in?

Reply:
So this is one place that webgpu is a bit underspecified. While upstream is a list of types, all the underlying apis seem to do this according to the order in the bitflags from LSB to MSB. This implementation also follows that behavior. If webgpu wants us to also swizzle the outputs, this is beyond the scope of the initial implementation as it requires a shader.

I have documented all this on the PipelineStatisticsTypes bitflag documentation.

Reply:
Yeah, the assumption was that a shader is needed regardless. If this no longer holds, we should reconsider the list there.